ADCS Firmware/src/main_adcs.cpp

```diff
@@ -1,21 +1,114 @@
 #include <Arduino.h>
 #include <Wire.h>
 #include "Board_settings.hpp"
+#include "DataTypes.hpp"
 
-void request_event()
+OperationMode op_mode = IDLE;
+GPSData gps_data;
+OrientationData o_data;
+RequestType req_type;
+
+void print_adcs_data(ADCSData *data)
+{
+    Serial.println("GPS Data");
+    Serial.println(data->gps_data.latitude);
+    Serial.println(data->gps_data.longutude);
+    
+    Serial.println("Orientaion Data");
+    Serial.println(data->o_data.abs_rot_x);
+    Serial.println(data->o_data.abs_rot_y);
+    Serial.println(data->o_data.abs_rot_z);
+    Serial.println(data->o_data.ang_rate_x);
+    Serial.println(data->o_data.ang_rate_y);
+    Serial.println(data->o_data.ang_rate_z);
+}
+
+void sample_data()
+{
+  // Update gps data and orientation data
+
+  // Randomise it for the sake of having something to send.
+  gps_data.latitude = random();
+  gps_data.longutude = random();
+
+  o_data.abs_rot_x = random();
+  o_data.abs_rot_y = random();
+  o_data.abs_rot_z = random();
+  o_data.ang_rate_x = random();
+  o_data.ang_rate_y = random();
+  o_data.ang_rate_z = random();
+
+  Serial.println("Sampled-------------------------");
+
+  return;
+}
+
+void send_data()
+{
+  Serial.println("Data requested.");
+  Serial.println((int)req_type);
+  switch (req_type)
+  {
+  case ADCS_DATA:
+    // Pack ADCS Data into buffer and
+    ADCSData adcs_data;
+    adcs_data.gps_data = gps_data;
+    adcs_data.o_data = o_data;
+    print_adcs_data(&adcs_data);
+    Serial.println("Writin ADCS data");
+    Wire.write(reinterpret_cast<uint8_t*>(&adcs_data), sizeof(adcs_data));
+    Serial.println("Wrote data :)");
+    break;
+
+  case STATUS:
+    // Pack status into buffer
+    Wire.write((int)op_mode);
+    break;
+  
+  default:
+    Serial.println("Unrecognised data send request type.");
+  }
+}
+
+void process_OBS_request(int n_bytes)
 {
-  Wire.write("hello!");
+  // while (Wire.available() < sizeof(OBSRequest)) {}; // Not the nicest way to wait but fine for now
+  Serial.println("Request incoming..");
+  // Take the first byte to see what type of request it is.
+  if (Wire.available()) {
+    req_type = (RequestType)Wire.read();
+    Serial.print("req_type set to ");
+    Serial.println((int) req_type);
+  }
+  
+  switch (req_type)
+  {
+  case MISSION_CHANGE:
+    // Grab the next byte to read the new mission type
+    if (Wire.available()) {
+      Serial.print("Changing op mode to ");
+      op_mode = (OperationMode)Wire.read();
+      Serial.println(op_mode);
+    }
```

Comment:
No response is sent to the OBC to report the actual mode status of the ADCS after a mode change request:

- [ ] add mode status reporting in accordance with the [data model specification](https://guorbit.github.io/obc-model/cloudview-1/html_export/output/obc-model/d24bab8168dc4ee9abd4ab6ed776b783.html#_JT8KcDEFEe6Qsa5LQcQyEA) and [existing code](https://github.com/guorbit/obc-firmware/blob/3a3812f4ec839a874c5e4509da34c217fcd0a180/obc-firmware/obc-firmware.asn)